### PR TITLE
Rename models Threshold, Graph to AlertThreshold, Chart #46

### LINKED ISSUE #46: [enhancement] Renaming the models Threshold and Graph in a backward compatible manner
Rename the models which are now called `Threshold`, `Graph` to `AlertSettings`, `Chart` respectively, in a backward compatible way (eg: provide a data migration which creates the alert table first, copies the data from the threshold table into alert and then remove the old threshold table, or an equivalently working alternative).

### PR DESCRIPTION
**Pull request by nepython**
I have tested it thoroughly on `Postgresql` multiple times both applying and unapplying the migration as can be seen below it works perfectly.
data_migration
**WARNING**: Still, please don't test on the main database to avoid any data loss (just in case and if there is any then please let me know along with the relational db name.)

I don't know why `Threshold` is to be renamed as `AlertSettings`. Since, the `s` is making it appear as `Alert Settingss` in the admin. :smile:  Let me know if I should just remove the extra `s` or the model should be renamed as `AlertSetting` only.

### CODE REVIEW

File: openwisp_monitoring/monitoring/migrations/0010_rename_threshold_alert_settings.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+                help_text='for how long should the alert_settings value be crossed before triggering an alert? The maximum allowed is 604800 seconds (7 days)',
+                validators=[MaxValueValidator(604800)],
+            ),
+        ),
+        migrations.AlterField(
+            model_name='alertsettings',
+            name='value',
+            field=models.FloatField(help_text='alert settings value'),
+        ),
+        migrations.RunPython(register_content_type, unregister_content_type),
```
Comment (nemesisdesign) [inline openwisp_monitoring/monitoring/migrations/0010_rename_threshold_alert_settings.py:47]: we had issues in the past when doing alters and inserts in the same transaction (one single migration file is executed in a single DB transaction), so move this to a separate migration file please.

File: README.rst [hunk 2]
```diff
@@ -23,12 +23,12 @@ Available Features
...
 * Displays ``Uptime``, ``Memory Status``, ``Load Status``, ``Addresses``, ``Neighbours`` of the **Device**
-* Displays ``Monitoring Graphs`` namely ``Uptime``, ``Packet Loss``, ``Round Trip Time``, ``Traffic`` in the form of lively charts
-* ``Monitoring Graphs`` can be viewed at resolutions of 1 day, 3 days, a week, a month and a year
+* Displays ``Monitoring Charts`` namely ``Uptime``, ``Packet Loss``, ``Round Trip Time``, ``Traffic`` in the form of lively charts
+* ``Monitoring Charts`` can be viewed at resolutions of 1 day, 3 days, a week, a month and a year
 * CSV Export of monitoring data
 * It allows you to add custom `Charts <https://github.com/openwisp/openwisp-monitoring/#openwisp_monitoring_charts>`_
-* It allows to create custom ``Metric``, select a ``Chart`` and set a ``Threshold``
-* Notifications are sent whenever ``Threshold`` value is crossed
+* It allows to create custom ``Metric``, select a ``Chart`` and set a ``AlertTheshold``
```
Comment (PabloCastellano) [inline README.rst:30]: typo: AlertThreshold

File: README.rst [hunk 3]
```diff
@@ -242,7 +242,7 @@ in terms of disk space.
 
 Whether ping checks are created automatically for devices.
 
-``OPENWISP_MONITORING_AUTO_GRAPHS``
+``OPENWISP_MONITORING_AUTO_CHARTS``
```
Comment (PabloCastellano) [inline README.rst:245]: Changing this setting name and updating to the latest version will add some friction to the upgrade. I would keep this name and if we want to keep consistency I would add a deprecation warning.

@nemesisdesign WDYT?

File: README.rst [hunk 4]
```diff
@@ -199,7 +199,7 @@ Everything is working normally.
 ~~~~~~~~~~~
 
 One of the metrics has a value which is not in the expected range
-(threshold value crossed).
+(AlertsSettings value crossed).
```
Comment (PabloCastellano) [inline README.rst:202]: AlertThreshold

File: README.rst [hunk 5]
```diff
@@ -262,7 +262,7 @@ Automatically created graphs.
 | **default**: | ``[{'key': 'ping', 'field_name': 'reachable'}]``                |
 +--------------+-----------------------------------------------------------------+
 
-Device metrics that are considered critical: when a threshold related to
+Device metrics that are considered critical: when a ``AlertTheshold`` related to
```
Comment (PabloCastellano) [inline README.rst:265]: AlertThreshold

File: README.rst [hunk 6]
```diff
@@ -647,9 +647,9 @@ Add the following to your ``settings.py``:
     # For extending check app
     CHECK_CHECK_MODEL = 'YOUR_MODULE_NAME.Check'
     # For extending monitoring app
-    MONITORING_GRAPH_MODEL = 'YOUR_MODULE_NAME.Graph'
+    MONITORING_CHART_MODEL = 'YOUR_MODULE_NAME.Chart'
     MONITORING_METRIC_MODEL = 'YOUR_MODULE_NAME.Metric'
-    MONITORING_THRESHOLD_MODEL = 'YOUR_MODULE_NAME.Threshold'
+    MONITORING_ALERTTHRESOLD_MODEL = 'YOUR_MODULE_NAME.AlertTheshold'
```
Comment (PabloCastellano) [inline README.rst:652]: AlertThreshold

File: README.rst [hunk 7]
```diff
@@ -709,12 +709,12 @@ Similarly for ``monitoring`` app, you can do it as:
...
 
-    from openwisp_monitoring.monitoring.admin import MetricAdmin, ThresholdAdmin
+    from openwisp_monitoring.monitoring.admin import MetricAdmin, AlertThesholdAdmin
 
     MetricAdmin.list_display.insert(1, 'my_custom_field')
     MetricAdmin.ordering = ['-my_custom_field']
-    ThresholdAdmin.list_display.insert(1, 'my_custom_field')
-    ThresholdAdmin.ordering = ['-my_custom_field']
+    AlertThesholdAdmin.list_display.insert(1, 'my_custom_field')
+    AlertThesholdAdmin.ordering = ['-my_custom_field']
```
Comment (PabloCastellano) [inline README.rst:717]: AlertThresholdAdmin

File: openwisp_monitoring/monitoring/models.py [hunk 8]
```diff
@@ -9,13 +9,14 @@ class Meta(AbstractMetric.Meta):
...
+        swappable = swappable_setting('monitoring', 'Chart')
 
 
-class Threshold(AbstractThreshold):
-    class Meta(AbstractThreshold.Meta):
+class AlertThreshold(AbstractAlertThreshold):
+    class Meta(AbstractAlertThreshold.Meta):
         abstract = False
-        swappable = swappable_setting('monitoring', 'Threshold')
+        db_table = 'monitoring_alertthreshold'
```
Comment (PabloCastellano) [inline openwisp_monitoring/monitoring/models.py:21]: I prefer `monitoring_alert_threshold` for readability but not a big deal

File: openwisp_monitoring/monitoring/tests/__init__.py [hunk 9]
```diff
@@ -53,21 +53,21 @@ def _create_object_metric(self, **kwargs):
             kwargs['is_healthy'] = True  # backward compatibility with old tests
         return self._create_general_metric(**opts)
 
-    def _create_threshold(self, **kwargs):
-        t = Threshold(**kwargs)
-        t.full_clean()
-        t.save()
-        return t
+    def _create_alert_threshold(self, **kwargs):
+        alert_s = AlertThreshold(**kwargs)
```
Comment (PabloCastellano) [inline openwisp_monitoring/monitoring/tests/__init__.py:57]: I understand alert_s comes from AlertSettings. Since we are not using that name anymore I think `t` or `at` are more meaningful.

File: openwisp_monitoring/monitoring/models.py (hunk 8 above)
Comment (nepython) [reply openwisp_monitoring/monitoring/models.py:21]: Django prefers to store and call it this way. For example the table names for `DeviceData` and `DeviceMonitoring` are `device_monitoring_devicedata` and `device_monitoring_devicemonitoring` respectively.

File: README.rst (hunk 3 above)
Comment (nepython) [reply README.rst:245]: @PabloCastellano, the GSoC project's name was `Improving OpenWISP Monitoring towards its first release`. Hence, I don't think it's been officially released yet but maybe it might be out for `beta` testing in which case I might be wrong :sweat_smile:

File: README.rst (hunk 2 above)
Comment (nepython) [reply README.rst]: Thanks for bringing out to my notice. I used a regex so a typo at one place means typo everywhere. It's corrected now.

File: README.rst (hunk 3 above)
Comment (PabloCastellano) [reply README.rst:245]: That's right. Let's wait for Federico who has more details on who is actually using this in production

Comment (nemesisdesign) [reply README.rst:245]: Thanks Pablo for the concern, in this case this change does not affect negatively already deployed systems so we can proceed.

File: openwisp_monitoring/monitoring/models.py (hunk 8 above)
Comment (nemesisdesign) [reply openwisp_monitoring/monitoring/models.py:21]: why is db_table needed here?

Comment (nepython) [reply openwisp_monitoring/monitoring/models.py:21]: @nemesisdesign,  django's `migraions.RenameModel` works great for a simple rename but it needs you to manually create the migration script and take care of few things in case there is some alteration of fields too([docs](https://docs.djangoproject.com/en/3.0/ref/migration-operations/#renamemodel)).

Thus, there were two slight changes I had to make. One is the above one and the other one is contenttype registration. I hope that explains :smile:

File: openwisp_monitoring/monitoring/base/models.py [hunk 10]
```diff
@@ -562,19 +564,19 @@ def _round(value, decimal_places):
...
+class AbstractAlertThreshold(TimeStampedEditableModel):
     _SECONDS_MAX = 60 * 60 * 24 * 7  # 7 days
     _SECONDS_HELP = (
-        'for how long should the threshold value be crossed before '
+        'for how long should the alert_threshold value be crossed before '
         'triggering an alert? The maximum allowed is {0} seconds '
         '({1} days)'.format(_SECONDS_MAX, int(_SECONDS_MAX / 60 / 60 / 24))
     )
-    _THRESHOLD_OPERATORS = (('<', _('less than')), ('>', _('greater than')))
+    _ALERTSETTINGS_OPERATORS = (('<', _('less than')), ('>', _('greater than')))
```
Comment (PabloCastellano) [inline openwisp_monitoring/monitoring/base/models.py:574]: alertsettings!

Comment (nepython) [reply openwisp_monitoring/monitoring/base/models.py]: Thanks!

**Comment by nepython** [comment]
The `qa-checks` are failing after the new release of openwisp-utils and thanks to it we have now identified minor issues which were otherwise unnoticed. I have opened #114 for the same.

File: README.rst (hunk 4 above)
Comment (PabloCastellano) [inline README.rst:202]: typo: AlertSettings

File: openwisp_monitoring/monitoring/migrations/0012_rename_graph_chart.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('monitoring', '0011_alert_settings_contenttype_registration'),
+    ]
+
+    operations = [
+        migrations.AlterModelTable(name='graph', table='monitoring_chart',),
+        migrations.RenameModel(old_name='Graph', new_name='Chart',),
+        migrations.AlterModelTable(name='chart', table=None,),
```
Comment (PabloCastellano) [inline openwisp_monitoring/monitoring/migrations/0012_rename_graph_chart.py:15]: Why is this needed?

File: openwisp_monitoring/monitoring/templates/admin/chart_inline.html [hunk 12]
```diff
@@ -21,11 +21,11 @@ <h3><b>{{ inline_admin_formset.opts.verbose_name|capfirst }}:</b>&nbsp;<span cla
   {% if inline_admin_form.needs_explicit_pk_field %}{{ inline_admin_form.pk_field.field }}{% endif %}
   {{ inline_admin_form.fk_field.field }}
   {% if inline_admin_form.original %}
-  <div class="form-row graph">
-      <div id="graph-{{ forloop.counter }}"></div>
+  <div class="form-row chart">
+      <div id="cahrt-{{ forloop.counter }}"></div>
```
Comment (PabloCastellano) [inline openwisp_monitoring/monitoring/templates/admin/chart_inline.html:25]: typo

File: openwisp_monitoring/monitoring/templates/admin/chart_inline.html [hunk 13]
```diff
@@ -21,11 +21,11 @@ <h3><b>{{ inline_admin_formset.opts.verbose_name|capfirst }}:</b>&nbsp;<span cla
...
       <script>
-          createGraph({{ inline_admin_form.form.instance.json|safe }}, null,
-                      'graph-{{ forloop.counter }}', null,
+          createChart({{ inline_admin_form.form.instance.json|safe }}, null,
+                      'cahrt-{{ forloop.counter }}', null,
```
Comment (PabloCastellano) [inline openwisp_monitoring/monitoring/templates/admin/chart_inline.html:28]: typo

File: openwisp_monitoring/monitoring/admin.py [hunk 14]
```diff
@@ -38,11 +38,15 @@ class MetricAdmin(TimeReadonlyAdminMixin, admin.ModelAdmin):
...
-@admin.register(Threshold)
-class ThresholdAdmin(TimeReadonlyAdminMixin, admin.ModelAdmin):
+@admin.register(AlertSettings)
+class AlertSettingsAdmin(TimeReadonlyAdminMixin, admin.ModelAdmin):
     list_display = ['metric', 'created', 'modified']
     search_fields = ['name']
     save_on_top = True
+
+    class Meta:
+        verbose_name = 'Alert Settings'
```
Comment (PabloCastellano) [inline openwisp_monitoring/monitoring/admin.py:51]: Should be translatable

File: openwisp_monitoring/monitoring/migrations/0010_rename_threshold_alert_settings.py [hunk 15]
```diff
@@ -0,0 +1,30 @@
+# Generated by Django 3.0.3 on 2020-05-30 19:12
```
Comment (PabloCastellano) [inline openwisp_monitoring/monitoring/migrations/0010_rename_threshold_alert_settings.py:1]: Please squash all the migrations unless there is a good reason not to

Comment (nepython) [reply openwisp_monitoring/monitoring/migrations/0010_rename_threshold_alert_settings.py:1]: This is the reason why I can't simply squash the migrations for this PR https://github.com/openwisp/openwisp-monitoring/pull/112#discussion_r433397418

File: openwisp_monitoring/monitoring/migrations/0012_rename_graph_chart.py (hunk 11 above)
Comment (nepython) [reply openwisp_monitoring/monitoring/migrations/0012_rename_graph_chart.py]: There were two options either doing this or the second one was to declare  `db_table` for the model when renaming. I think declaring `db_table` in the `Meta` would be a better thing to do here (though I am unsure what is the general practice)

File: openwisp_monitoring/monitoring/migrations/0010_rename_threshold_alert_settings.py (hunk 15 above)
Comment (PabloCastellano) [reply openwisp_monitoring/monitoring/migrations/0010_rename_threshold_alert_settings.py:1]: fair enough! :+1:

File: openwisp_monitoring/monitoring/migrations/0012_rename_graph_chart.py (hunk 11 above)
Comment (PabloCastellano) [reply openwisp_monitoring/monitoring/migrations/0012_rename_graph_chart.py]: I see you have now removed `migrations.AlterModelTable(name='chart', table=None,),` :+1:

File: openwisp_monitoring/monitoring/migrations/0011_alert_settings_contenttype_registration.py [hunk 16]
```diff
@@ -0,0 +1,30 @@
+# Generated by Django 3.0.3 on 2020-06-02 14:12
+
+from django.contrib.contenttypes.models import ContentType
+from django.core.exceptions import ObjectDoesNotExist
+from django.db import migrations
+
+
+def register_content_type(apps, schema_editor):
+    """Registers AlertThreshold if not yet registered"""
```
Comment (PabloCastellano) [inline openwisp_monitoring/monitoring/migrations/0011_alert_settings_contenttype_registration.py:9]: AlertSettings

File: tests/openwisp2/sample_monitoring/migrations/0004_alert_settings_contenttype_registration.py [hunk 17]
```diff
@@ -0,0 +1,30 @@
+# Generated by Django 3.0.3 on 2020-06-02 14:15
+
+from django.contrib.contenttypes.models import ContentType
+from django.core.exceptions import ObjectDoesNotExist
+from django.db import migrations
+
+
+def register_content_type(apps, schema_editor):
+    """Registers AlertThreshold if not yet registered"""
```
Comment (PabloCastellano) [inline tests/openwisp2/sample_monitoring/migrations/0004_alert_settings_contenttype_registration.py:9]: AlertSettings

**Comment by nepython** [comment]
Done :+1:
It  is getting slightly confusing now between AlertThreshold and AlertSettings :sweat_smile:

File: openwisp_monitoring/monitoring/models.py [hunk 18]
```diff
@@ -9,13 +9,15 @@ class Meta(AbstractMetric.Meta):
         swappable = swappable_setting('monitoring', 'Metric')
 
 
-class Graph(AbstractGraph):
-    class Meta(AbstractGraph.Meta):
+class Chart(AbstractChart):
+    class Meta(AbstractChart.Meta):
         abstract = False
-        swappable = swappable_setting('monitoring', 'Graph')
+        db_table = 'monitoring_chart'
```
Comment (nemesisdesign) [inline openwisp_monitoring/monitoring/models.py:15]: can you explain why setting the db_table explicitly here and in the other occurrences is needed? Can we do it without doing this?

Comment (nepython) [reply openwisp_monitoring/monitoring/models.py:15]: Django's migrations.RenameModel works great for a simple rename but it needs you to manually create the migration script and take care of few small things like renaming the table name (not much trouble and better than deleting and creating a new model, I think)
>Can we do it without doing this?

Yes, we can do the same in two approaches one is using `db_table` while other is adding a line in the migrations script. I did it some time back but was asked if it can be done without this so shifted to `db_table` https://github.com/openwisp/openwisp-monitoring/pull/112/commits/f7ea2de262133fa7e16ce424f6787dd959c0a77b#diff-5f2ccbbd9d955a5d9006eb87270c59d3L15.
I am not sure which approach is better here, please let me know.

Comment (nemesisdesign) [reply openwisp_monitoring/monitoring/models.py:15]: It's ok to have additional lines in the migration scripts.
`db_table` in the model options (`Meta` class) should be used only when dealing with legacy systems.

Comment (nepython) [reply openwisp_monitoring/monitoring/models.py]: Done :+1:
Can you please explain why it's reserved specifically for legacy systems, I didn't came across this earlier so might learn something. Thanks.

File: openwisp_monitoring/device/templates/admin/config/device/change_form.html [hunk 19]
```diff
@@ -312,66 +312,66 @@ <h2>{% trans 'Neighbors' %}</h2>
 </div>
 {% endif %}
 {% if not add %}
-<div class="inline-group" id="ow-graph-container">
+<div class="inline-group" id="ow-chart-container">
     <fieldset class="module">
-        <h2>{% trans 'Monitoring Graphs' %}</h2>
-        <div id="ow-graph-time">
+        <h2>{% trans 'Monitoring Charts' %}</h2>
```
Comment (nemesisdesign) [inline openwisp_monitoring/device/templates/admin/config/device/change_form.html:317]: Let's shorten this to just `Charts`, I'm doing the same for the other tabs (they're too long).

File: openwisp_monitoring/monitoring/migrations/0012_rename_graph_chart.py (hunk 11 above)
Comment (PabloCastellano) [inline openwisp_monitoring/monitoring/migrations/0012_rename_graph_chart.py:15]: @nepython I see you removed db_table but came back to this `migrations.AlterModelTable(name='chart', table=None,)`. Why is this necessary?

Comment (PabloCastellano) [reply openwisp_monitoring/monitoring/migrations/0012_rename_graph_chart.py:15]: I quote what I said on gitter last week or so:

> In any case, I don't see any problem if the table in the DB is not renamed, after all we are using an ORM that abstracts these details

We can squash all migrations before releasing the first version of openwisp-monitoring if necessary

Comment (PabloCastellano) [reply openwisp_monitoring/monitoring/migrations/0012_rename_graph_chart.py:15]: Last thoughts:

This issue was only about renaming the models. This PR contains also changes for the db schema, the css classes and some other extra changes. I'm not saying it's wrong but must stick to the issue description in general and open new issues and discuss if the scope gets broader, otherwise the changes are by default out of the scope of the issue.

Comment (nemesisdesign) [reply openwisp_monitoring/monitoring/migrations/0012_rename_graph_chart.py:15]: The renaming of other occurrences of graph is ok, I think it's consistent with the issue description.
I also don't understand very well why `table=None` is needed, @nepython can you please explain?

Comment (nepython) [reply openwisp_monitoring/monitoring/migrations/0012_rename_graph_chart.py:15]: I don't know for sure but AlterModelTable refers to the `db_table` in Meta class. Since, it is not declared, Django generates this migration but I have checked the database there were no changes actually. `table=None` means restore the default name of the model table which in this case happens to be the same as we have renamed two lines above. I tried without these as @PabloCastellano suggested no renaming at all but ended up losing data and there were some other issues cropping up when the table wasn't being renamed.

File: README.rst [hunk 20]
```diff
@@ -23,11 +23,11 @@ Available Features
...
 
 * Displays ``Uptime``, ``Memory Status``, ``Load Status``, ``Addresses``, ``Neighbours`` of the **Device**
-* Displays ``Monitoring Graphs`` namely ``Uptime``, ``Packet Loss``, ``Round Trip Time``, ``Traffic`` in the form of lively charts
-* ``Monitoring Graphs`` can be viewed at resolutions of 1 day, 3 days, a week, a month and a year
+* Displays ``Charts`` namely ``Uptime``, ``Packet Loss``, ``Round Trip Time``, ``Traffic`` in the form of lively charts
+* ``Charts`` can be viewed at resolutions of 1 day, 3 days, a week, a month and a year
 * CSV Export of monitoring data
 * It allows you to add custom `Charts <https://github.com/openwisp/openwisp-monitoring/#openwisp_monitoring_charts>`_
-* It allows to create custom ``Metric``, select a ``Chart`` and set a ``Threshold``
+* It allows to create custom ``Metric``, select a ``Chart`` and set a ``AlertSettings``
```
Comment (nemesisdesign) [inline README.rst:30]: let's try to write documentation that is not only aimed at developers, so we don't necessarily have to refer to the model names, we should try to use more human readable names.

File: README.rst [hunk 21]
```diff
@@ -199,7 +199,7 @@ Everything is working normally.
 ~~~~~~~~~~~
 
 One of the metrics has a value which is not in the expected range
-(threshold value crossed).
+(AlertSettings value crossed).
```
Comment (nemesisdesign) [inline README.rst:202]: applies to here as well, see how I have edited it in the branch where this will be merged

File: README.rst [hunk 22]
```diff
@@ -208,7 +208,7 @@ Example: CPU usage should be less than 90% but current value is at 95%.
 
 One of the metrics defined in ``OPENWISP_MONITORING_CRITICAL_DEVICE_METRICS``
 has a value which is not in the expected range
-(threshold value crossed).
+(AlertSettings value crossed).
```
Comment (nemesisdesign) [inline README.rst:212]: applies to here as well, see how I have edited it in the branch where this will be merged

File: README.rst [hunk 23]
```diff
@@ -428,13 +428,13 @@ it is not sent if the response was not successful.
...
-| Notification Types    | Uses                                                                         |
-+-----------------------+------------------------------------------------------------------------------+
-| ``threshold_crossed`` | Used for creating notification when a metric exceeds threshold value.        |
-+-----------------------+------------------------------------------------------------------------------+
-| ``under_threshold``   | Used for creating notification when a metric falls behind threshold value.   |
-+-----------------------+------------------------------------------------------------------------------+
++----------------------------+---------------------------------------------------------------------------------+
+|    Notification Types      | Uses                                                                            |
++----------------------------+---------------------------------------------------------------------------------+
+| ``alert_settings_crossed`` | Used for creating notification when a metric exceeds alert settings value.      |
```
Comment (nemesisdesign) [inline README.rst:434]: AlertSettings.value is still indicated as the threshold value so I think we can keep using this form because is clear, I will rename this.

File: README.rst [hunk 24]
```diff
@@ -428,13 +428,13 @@ it is not sent if the response was not successful.
...
++----------------------------+---------------------------------------------------------------------------------+
+|  ``under_alert_settings``  | Used for creating notification when a metric falls behind alert settings value. |
```
Comment (nemesisdesign) [inline README.rst:436]: I noticed this only now, I don't like very much this name, I will change it (this is for you @pandafy)
